server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -151,32 +161,42 @@ protected void registerWithDeprecatedHandler(RestRequest.Method method, String p
      * @param handler The handler to actually execute
      * @param method GET, POST, etc.
      */
-    protected void registerHandler(RestRequest.Method method, String path, RestHandler handler) {
+    protected void registerHandler(Method method, String path, RestApiCompatibleVersion version, RestHandler handler) {
         if (handler instanceof BaseRestHandler) {
             usageService.addRestHandler((BaseRestHandler) handler);
         }
-        registerHandlerNoWrap(method, path, handlerWrapper.apply(handler));
+        registerHandlerNoWrap(method, path, version, handlerWrapper.apply(handler));
     }
 
-    private void registerHandlerNoWrap(RestRequest.Method method, String path, RestHandler maybeWrappedHandler) {
-        final RestApiCompatibleVersion version = maybeWrappedHandler.compatibleWithVersion();
+    private void registerHandlerNoWrap(Method method, String path, RestApiCompatibleVersion version, RestHandler handler) {
         assert RestApiCompatibleVersion.minimumSupported() == version || RestApiCompatibleVersion.currentVersion() == version
             : "REST API compatibility is only supported for version " + RestApiCompatibleVersion.minimumSupported().major;
 
-        handlers.insertOrUpdate(path, new MethodHandlers(path, maybeWrappedHandler, method),
-            (mHandlers, newMHandler) -> mHandlers.addMethods(maybeWrappedHandler, method));
+        handlers.insertOrUpdate(path,
+            new MethodHandlers(path).addMethod(method, version, handler),
+            (handlers, ignoredHandler) -> handlers.addMethod(method, version, handler));
+    }
+
+    public void registerHandler(final Route route, final RestHandler handler) {
+        if (route.isReplacement()) {
+            Route replaced = route.getReplacedRoute();
+            registerAsReplacedHandler(route.getMethod(), route.getPath(), RestApiCompatibleVersion.currentVersion(), handler,
+                replaced.getMethod(), replaced.getPath(), route.getRestApiCompatibleVersion());
+        } else if (route.isDeprecated()) {
+            registerAsDeprecatedHandler(route.getMethod(), route.getPath(), route.getRestApiCompatibleVersion(), handler,
+                route.getDeprecationMessage());
+        } else {
+            // it's just a normal route
+            registerHandler(route.getMethod(), route.getPath(), RestApiCompatibleVersion.currentVersion(), handler);
```

Comment:
so what if a handler was not deprecated previously? The deprecation logging was just done within the RestHandler itself (often scenario in types removal)
Shouldn't we get a `RestApiCompatibleVersion` from a Route in this case?

Reply:
Let's catch up on this one over slack or zoom tomorrow.